File: frontend/src/api/enrichment.ts
import { api } from "./client"
import type { EnrichmentJob, LibraryFilters } from "@/types/domain"
import { filtersToApiBody } from "./schools"

export async function startEnrichmentJob(schoolIds: number[]): Promise<EnrichmentJob> {
  return api.post<EnrichmentJob>("/enrichment-jobs", { school_ids: schoolIds })
}

/** Enrich every school matching a Library filter set (highest score first),
 * optionally capped at `limit` -- the filter-based counterpart to
 * pullIntoPipeline, so "enrich all N matching" isn't bounded by page size. */
export async function startEnrichmentJobFromFilters(
  filters: LibraryFilters,
  limit: number | null
): Promise<EnrichmentJob> {
  return api.post<EnrichmentJob>("/enrichment-jobs", {
    filters: filtersToApiBody(filters),
    limit,
  })
}

/** All jobs, most recent first -- includes completed ones so the tray can
 * show a "done" summary before the user dismisses it (a job that's
 * already finished would otherwise never appear in a status-filtered
 * query). */
export async function listRecentJobs(): Promise<EnrichmentJob[]> {
  return api.get<EnrichmentJob[]>("/enrichment-jobs")
}

export async function getJob(id: number): Promise<EnrichmentJob> {
  return api.get<EnrichmentJob>(`/enrichment-jobs/${id}`)
}

/** Stops a pending/running job before its next school -- the one already
 * mid-scrape still finishes, everything after it is skipped. */
export async function cancelEnrichmentJob(id: number): Promise<EnrichmentJob> {
  return api.post<EnrichmentJob>(`/enrichment-jobs/${id}/cancel`, {})
}

File: src/levelup/api/v1/enrichment.py
from __future__ import annotations

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException
from sqlalchemy.orm import Session, joinedload

from levelup.api.v1.schemas import EnrichmentJobOut, EnrichmentJobRequest
from levelup.api.v1.schools import _apply_filters
from levelup.core.db import get_session
from levelup.core.security import get_current_user
from levelup.models.enrichment import EnrichmentJob, EnrichmentJobItem
from levelup.models.school import School
from levelup.models.score import CurrentScore, SchoolScore
from levelup.models.user import User
from levelup.services.enrichment.jobs import cancel_job, create_job, run_job

router = APIRouter(prefix="/enrichment-jobs", tags=["enrichment"])


@router.post("", response_model=EnrichmentJobOut, status_code=202)
def start_enrichment_job(
    body: EnrichmentJobRequest,
    background_tasks: BackgroundTasks,
    session: Session = Depends(get_session),
    user: User = Depends(get_current_user),
):
    # Mirror /pipeline/pull: accept either an explicit id list (checkbox
    # selection) or a filter set, so a whole filtered Library segment can be
    # enriched at once -- highest-score-first, capped by an optional limit --
    # rather than being bounded by whatever fit on the current page.
    if body.school_ids is not None:
        school_ids = body.school_ids
    elif body.filters is not None:
        query = _apply_filters(
            session.query(School.id)
            .outerjoin(CurrentScore, CurrentScore.school_id == School.id)
            .outerjoin(SchoolScore, SchoolScore.id == CurrentScore.score_id),
            **body.filters,
        )
        query = query.order_by(SchoolScore.total_score.desc().nulls_last())
        if body.limit is not None:
            query = query.limit(body.limit)
        school_ids = [row[0] for row in query.all()]
    else:
        raise HTTPException(400, "Provide either school_ids or filters")

    if not school_ids:
        raise HTTPException(400, "No schools matched -- nothing to enrich")

    job = create_job(session, school_ids, requested_by=user.id)
    background_tasks.add_task(run_job, job.id)
    return job


@router.get("", response_model=list[EnrichmentJobOut])
def list_jobs(session: Session = Depends(get_session), status: str | None = None):
    query = session.query(EnrichmentJob).options(joinedload(EnrichmentJob.items).joinedload(EnrichmentJobItem.school))
    if status:
        query = query.filter(EnrichmentJob.status == status)
    return query.order_by(EnrichmentJob.requested_at.desc()).all()


@router.get("/{job_id}", response_model=EnrichmentJobOut)
def get_job(job_id: int, session: Session = Depends(get_session)):
    return (
        session.query(EnrichmentJob)
        .options(joinedload(EnrichmentJob.items).joinedload(EnrichmentJobItem.school))
        .filter_by(id=job_id)
        .one()
    )


@router.post("/{job_id}/cancel", response_model=EnrichmentJobOut)
def cancel_enrichment_job(job_id: int, session: Session = Depends(get_session)):
    cancel_job(session, job_id)
    return (
        session.query(EnrichmentJob)
        .options(joinedload(EnrichmentJob.items).joinedload(EnrichmentJobItem.school))
        .filter_by(id=job_id)
        .one()
    )
